ipq807x: add initial support for prpl Foundation Haze board

ynezz opened the pull request:
Haze is prpl Foundation's reference board (WNC LVRP).

Board info:
* IPQ8072A SoC
* 2 GiB RAM
* 4 GiB eMMC
* 8MiB SPI NOR (MX25U6435F)
* 3x 1GigE ports (QCA8075)
* 1x 10GigE port (AQR113C)
* 1x SFP cage
* WiFi 6GHz 160MHz (QCN9074)
* WiFi 5GHz 80+80MHz (QCN5054)
* WiFi 2.4G (QCN5024)
* ARM Standard 20-pin 2.54mm/0.1" JTAG (1V8 !!!)
* Bluetooth v5.0 + EDR with integrated Class 1 PA (CYW20704)
* 1x M.2 B-key socket with PCIe 3.0
* 1x USB 3.0 port
* UART 4-pin 3V3(arrow),RX,TX,GND (115200 8N1)
* Reset and WPS buttons

john-tho committed:
rtl8812au-ct: bump to fix kernel 6.1 compile

 9b2b203 fix usb_recv_tasklet -Wcast-function-type
 a027da5 fix kernel 6.1 80211 link_id
 7a9c802 fix build for kernel 6.1 prandom
 3a3eb24 fix build for kernel 5.17 PDE_DATA
 fe2afbd fix build for kernel 5.17 const netdev->dev_addr
 7275bae fix build for 5.17 kernel complete_and_exit
 c9c2aa7 Update usb_intf.c

revert a027da5, the 6.1 link_id changes patch so that our
linux 5.15 + 80211 6.1 backports tree still compiles
with our downstream api_update patch.

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -0,0 +1,311 @@
+
+	mmc-hs400-1_8v;
+	mmc-hs200-1_8v;
+	mmc-ddr-1_8v;
```
No need to add the feature advertisements as these exact ones are present in the SoC DTSI already.

ynezz replied:
Ok, will check that.

ynezz replied:
Removed.

robimarko commented on target/linux/ipq807x/files/arch/arm64/boot/dts/qcom/ipq8072-haze.dts:
```diff
@@ -0,0 +1,311 @@
+		spi-max-frequency = <50000000>;
+
+		partitions {
+			compatible = "qcom,smem-part";
```
Are you sure that the SMEM parser is working?

Cause, I am seeing that you are using uppercase partition names in the caldata and u-boot scripts, but SMEM parser normalizes everything to lowercase

ynezz replied:
> Are you sure that the SMEM parser is working?

I hope, that it would be working.

I got devices which are booting from eMMC, so the SPI flash is not initialized with smem partition structure, but I've assumed, that it's quite common format, so it would work out of the box in the future. 

It wont probably hurt, but I'm fine with removing it as well.

> Cause, I am seeing that you are using uppercase partition names in the caldata and u-boot scripts

Thanks for double checking, but those are correct, they're eMMC partitions. I've double checked sha256sums of the extracted caldata binaries from QSDK and OpenWrt, they're same.

robimarko replied:
Ok, that makes sense.

robimarko commented on target/linux/ipq807x/base-files/etc/hotplug.d/firmware/11-ath11k-caldata:
```diff
@@ -22,10 +22,17 @@ case "$FIRMWARE" in
 		caldata_extract "0:art" 0x1000 0x20000
 		;;
+	prpl,haze)
+                caldata_extract_mmc "0:ART" 0x1000 0x20000
```
Where is the caldata stored, on NOR or eMMC?

ynezz replied:
its stored on eMMC

robimarko commented on target/linux/ipq807x/image/generic.mk:
```diff
@@ -108,6 +108,18 @@ define Device/netgear_wax218
+	DEVICE_DTS_CONFIG := config@hk09
+	SOC := ipq8072
+	DEVICE_PACKAGES += ath11k-firmware-qcn9074 ipq-wifi-prpl_haze kmod-ath11k-pci
+	DEFAULT_PACKAGES += ath11k-firmware-qcn9074 ipq-wifi-prpl_haze kmod-ath11k-pci
```
Why are you extending the target `DEFAULT_PACKAGES`?

ynezz replied:
To have those packages right in the initramfs, so can test automatically wireless part in the [daily runtime CI testing](https://gitlab.com/ynezz/openwrt-device-runtime-testing/-/pipelines/888312563/builds) without any additional steps. I plan to replace current ipq807x DUT Dynalink WRX36 with Haze as it have more features.

robimarko replied:
Is this working for you?
For me, it does not expand the default package list.
Also, this would mean that all boards would have to carry these and I dont think that is the desired effect.

ynezz replied:
It doesn't work as expected, indeed.

robimarko commented:
You can already submit the BDF-s, no need to wait for that, @Ansuel can also update the regulatory data in them to the current one.

ynezz commented:
> You can already submit the BDF-s, no need to wait for that

I hope, that I did, is this fine https://github.com/openwrt/firmware_qca-wireless/pull/8 ?

> can also update the regulatory data in them to the current one.

Nice, out of curiosity, how is that being done?

> M.2 should be just working as long as kmods for the end device, nvme and like are included.
> As far as SFP goes, that I suspect won't just work currently but I have no devices with the SFP cage.

I simply wanted to make it clear what is tested and expected to be working out of the box right now, will tackle the missing features later, unless someone else beats me to it. 

> Is this board obtainable?

Good question, as far as I'm aware, its currently available to prpl Foundation members in very low quantities, but I'll ask around and will let you know.

robimarko commented:
BDF-s have been merged, so just bump the ipq-wifi sources.

Well, we pulled the latest regulatory info and Ansuel has a script to update the binary BDF-s.

Ansuel commented:
Updated the BDF with new regDB

mrkiko commented:
Hi!
May you add some installation/debrick instructions?

Thanks!

ynezz committed:
ipq-wifi: update to version 2023-06-03

Contains following updates:

 * ipq8074: update RegDB in new submitted BDF
 * Revert "ipq8074: update RegDB in new submitted BDF"
 * qcn9074: update RegDB in new submitted BDF
 * ipq8074: update RegDB in new submitted BDF
 * qca-wireless: ipq40xx: add BDFs for ZTE MF287+
 * Add BDFs for prpl Foundation Haze board

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz committed:
ipq807x: add initial support for prpl Foundation Haze board

Haze is prpl Foundation's reference board (WNC LVRP).

Board info:

 - IPQ8072A SoC
 - 2 GiB RAM
 - 4 GiB eMMC
 - 8MiB SPI NOR (MX25U6435F)

 - 3x 1GigE ports (QCA8075)
 - 1x 10GigE port (AQR113C)
 - 1x SFP cage

 - WiFi 6GHz 160MHz (QCN9074)
 - WiFi 5GHz 80+80MHz (QCN5054)
 - WiFi 2.4G (QCN5024)

 - ARM Standard 20-pin 2.54mm/0.1" JTAG (1V8 !!!)
 - Bluetooth v5.0 + EDR with integrated Class 1 PA (CYW20704)
 - 1x M.2 B-key socket with PCIe 3.0
 - 1x USB 3.0 port
 - UART marked J6 is 4-pin 2.54mm/0.1" connector 3V3(arrow),RX,TX,GND (115200 8N1)
 - Reset and WPS buttons

Flashing instructions:

 1. From U-Boot boot OpenWrt using initramfs image:

    IPQ807x# tftpboot openwrt-ipq807x-generic-prpl_haze-initramfs-uImage.itb && bootm

 2. In OpenWrt running from initramfs execute sysupgrade:

    root@OpenWrt:/# sysupgrade -n /tmp/openwrt-ipq807x-generic-prpl_haze-squashfs-sysupgrade.bin

Work in progress/known issues:

 * SFP feature not implemented/tested
 * M.2 feature not implemented/tested
 * Bluetooth feature not implemented/tested
 * 6GHz wireless should be working, but not tested
 * MAC address assigments for LAN interfaces

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz commented:
>  May you add some installation/debrick instructions?

@mrkiko those are included in the commit message

ynezz committed:
ipq807x: image: fix eMMC flashing/recovery from within initramfs

Having initramfs image built with same config as on buildbots:

 CONFIG_TARGET_MULTI_PROFILE=y
 CONFIG_TARGET_ALL_PROFILES=y
 CONFIG_TARGET_PER_DEVICE_ROOTFS=y

Its currently impossible to flash/recover the device using that image as
losetup is missing:

 root@OpenWrt:/# sysupgrade -v /tmp/openwrt-ipq807x-generic-prpl_haze-squashfs-sysupgrade.bin
 ...
 /lib/upgrade/do_stage2: line 38: losetup: not found
 Failed to detach all loop devices. Skip this try.

So lets fix it by including the needed utils for sysupgrade in
DEFAULT_PACKAGES set.

Signed-off-by: Petr Štetiar <ynezz@true.cz>

ynezz commented:
Pulled in #12746 to fix that CI build failure.

ynezz commented:
Merged in 07fe8bc62a86
